### PR TITLE
Split installation into multiple steps

### PR DESCRIPTION
**Pull request by matthiasdiener**

### CODE REVIEW

**Comment by matthiasdiener** [comment]
@majosm: Does this fix the issue you mentioned in #6?

**Comment by majosm** [comment]
Did you mean to use `have_conda` in the if statement? It looks like it's still checking for the directory.

**Comment by matthiasdiener** [comment]
> Did you mean to use `have_conda` in the if statement? It looks like it's still checking for the directory.

No, the check for the `miniconda3` directory. the have_conda check does not work as intended, as the CI VMs already have a conda version installed, but which does not install pocl correctly...

**Comment by majosm** [comment]
Oh, hmm. Is there maybe a way to test whether the `conda` present is one that installs pocl correctly or not? Checking for `~/miniconda3` isn't necessarily general (if this is the goal) as it can be installed elsewhere. For example, on my Mac it's installed in `/opt/miniconda3`, and on Quartz I put it in `~/.miniconda3` (just to be difficult :P).

**Comment by matthiasdiener** [comment]
@majosm @MTCam @inducer  Please take a look at the current version. The pip/conda installation is split into 2 separate scripts now, and conda installation should be more robust (= fewer segfaults 😂)

**Comment by matthiasdiener** [comment]
This now works on Power9 and is able to run wave-eager with POCL on the CPUs.

**Comment by majosm** [comment]
Any chance we could split the script up between the `conda init` and the environment creation + dependency installation? I guess what I originally had in mind was something along the lines of:

```
# Run if you don't already have a suitable conda install
./get-conda.sh 

# Run if you don't already have an environment you want to install into
conda create -n <name>
conda activate <name>

# Install pocl and pip stuff
./get-dependencies.sh
```

The above code could be wrapped up into a single `install.sh` script for noobs who don't have anything installed.

**Comment by matthiasdiener** [comment]
> Any chance we could split the script up between the `conda init` and the environment creation + dependency installation? I guess what I originally had in mind was something along the lines of:
> 
> ```
> # Run if you don't already have a suitable conda install
> ./get-conda.sh 
> 
> # Run if you don't already have an environment you want to install into
> conda create -n <name>
> conda activate <name>
> 
> # Install pocl and pip stuff
> ./get-dependencies.sh
> ```
> 
> The above code could be wrapped up into a single `install.sh` script for noobs who don't have anything installed.

So 2 scripts now become 3 scripts? This is becoming Inception-level scripting 🤣

**Comment by majosm** [comment]
https://en.wikipedia.org/wiki/Fundamental_theorem_of_software_engineering :wink:

**Comment by matthiasdiener** [comment]
> > Any chance we could split the script up between the `conda init` and the environment creation + dependency installation? I guess what I originally had in mind was something along the lines of:
> > ```
> > # Run if you don't already have a suitable conda install
> > ./get-conda.sh 
> > 
> > # Run if you don't already have an environment you want to install into
> > conda create -n <name>
> > conda activate <name>
> > 
> > # Install pocl and pip stuff
> > ./get-dependencies.sh
> > ```
> > 
> > 
> > The above code could be wrapped up into a single `install.sh` script for noobs who don't have anything installed.
> 
> So 2 scripts now become 3 scripts? This is becoming Inception-level scripting 🤣

Done.

File: install-dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        echo "==== ERROR: incomplete git clone. Please run:"
+        echo "====   git submodule init && git submodule update"
+        echo "==== to fetch emirge's submodules."
+        exit 1
+    fi
+done
+
+echo "==== Installing Conda packages"
+
+conda update --all --yes
```
Comment (majosm) [inline install-dependencies.sh:18]: Can the `conda update` be left out, or is it needed here?

File: install-spack.sh [hunk 2]
```diff
@@ -35,15 +28,10 @@ pocl_dir=$(spack find -p pocl| tail -1 | awk '{print $2}')
 export OCL_ICD_VENDORS=$pocl_dir/etc/OpenCL/vendors/
 echo 'export OCL_ICD_VENDORS=$pocl_dir/etc/OpenCL/vendors/' >> $HOME/.bashrc
 
-echo "==== Installing pip packages"
-pip install wheel pyvisfile numpy
-
-for module in dagrt leap loopy meshmode grudge mirgecom; do (cd $module && pip install -e .) ; done
-EOF
-
+./install-pip.sh
```
Comment (majosm) [inline install-spack.sh:31]: `install-pip.sh` no longer exists. :) Though maybe it should? And be called by both `install-spack.sh` and `install-dependencies.sh`? (More scripts, yay!)

File: install.sh [hunk 3]
```diff
@@ -7,43 +7,22 @@ echo "# This script installs the dependencies for emirge. #"
...
-myarch=$(uname -m)
+export PATH=$MY_CONDA_DIR/bin:$PATH
 
-if [[ ! -d ~/miniforge3 ]]; then
-    echo "Installing Miniforge"
-    wget -c --quiet https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$myos-$myarch.sh
-    bash Miniforge3-$myos-$myarch.sh -b
-fi
+echo "==== Create 'dgfem' conda environment"
+conda init
```
Comment (majosm) [inline install.sh:17]: Should `conda init` go here or in `install-conda.sh`? (Not really sure what makes more sense.)

File: install.sh [hunk 4]
```diff
@@ -7,43 +7,22 @@ echo "# This script installs the dependencies for emirge. #"
...
+conda create --name dgfem --yes
 
-export CONDA=$HOME/miniforge3
-export PATH=$HOME/miniforge3/bin:$PATH
-export OCL_ICD_VENDORS=$HOME/miniforge3/etc/OpenCL/vendors/
-echo 'export OCL_ICD_VENDORS=$HOME/miniforge3/etc/OpenCL/vendors/' >> $HOME/.bashrc
+$MY_CONDA_DIR/bin/activate dgfem
```
Comment (majosm) [inline install.sh:20]: Did `conda activate dgfem` give you an error here? If so, I think you can add `eval "$(conda shell.bash hook)"` before it.

File: install.sh [hunk 5]
```diff
@@ -7,43 +7,22 @@ echo "# This script installs the dependencies for emirge. #"
...
-	exit 1
-fi
+export MY_CONDA_DIR=$HOME/miniforge3
 
-myos=$(uname)
-[[ $myos == "Darwin" ]] && myos="MacOSX"
+./install-conda.sh
 
-myarch=$(uname -m)
+export PATH=$MY_CONDA_DIR/bin:$PATH
```
Comment (majosm) [inline install.sh:14]: Tempted to say this should go in `install-conda.sh`.

File: install-dependencies.sh (hunk 1 above)
Comment (matthiasdiener) [reply install-dependencies.sh:18]: I think it's better to leave it here. Even when downloading the latest Miniforge install script, there are always updates immediately available.

File: install-spack.sh (hunk 2 above)
Comment (matthiasdiener) [reply install-spack.sh:31]: Good catch. I forgot to rename install-pip.sh to install-dependencies.sh here.

File: install.sh (hunk 3 above)
Comment (matthiasdiener) [reply install.sh:17]: I think it's fine here.

File: install.sh (hunk 5 above)
Comment (matthiasdiener) [reply install.sh:14]: You can't export variables from a sub-script to its calling script.

**Comment by matthiasdiener** [comment]
> https://en.wikipedia.org/wiki/Fundamental_theorem_of_software_engineering 😉

When writing scripts, I personally follow more the Spice Girls school of development: https://en.wikipedia.org/wiki/2_Become_1

**Comment by matthiasdiener** [comment]
> Added some thoughts. Also, one more: now that we have an abundance of scripts (whose fault is that? 😅) it might be good for clarity to rename or move some of them. Thinking we should just have the "install everything" scripts for conda and spack in the top level directory and throw the sub-pieces in a directory somewhere (`scripts`?).

We could do that, but I think then people would start complaining about not finding the right script 😅. We could always do that in a future PR.

File: install.sh (hunk 4 above)
Comment (majosm) [reply install.sh:20]: Was thinking you might be able to get rid of the `MY_CONDA_DIR` definition, but it looks like you need it for the `PATH` stuff anyway.

File: install-dependencies.sh (hunk 1 above)
Comment (majosm) [reply install-dependencies.sh]: Could we maybe put it in `install-conda.sh`? Just thinking that if I'm installing emirge into an existing environment, I may not want to update every package that I have (just whatever is needed for mirgecom to work).

Comment (matthiasdiener) [reply install-dependencies.sh]: Good point. Fixed in the most recent commit.

**Comment by majosm** [comment]
> > Added some thoughts. Also, one more: now that we have an abundance of scripts (whose fault is that? 😅) it might be good for clarity to rename or move some of them. Thinking we should just have the "install everything" scripts for conda and spack in the top level directory and throw the sub-pieces in a directory somewhere (`scripts`?).
> 
> We could do that, but I think then people would start complaining about not finding the right script 😅. We could always do that in a future PR.

Yeah, I was just trying to think how to make it clearer which scripts do what. For example, if I didn't know better I would find it confusing that `install-conda.sh` and `install-spack.sh` aren't equivalent (i.e. I would think that one installs everything with conda and the other installs everything with spack). Putting sub-scripts in a directory would be one way to fix that, but there may be alternatives.

Otherwise, looks good to me. Not sure if anyone else wants to review before merging? I'll leave it open for the moment.

**Pull request merged**